### PR TITLE
[WIP] Implement drag drop feature

### PR DESCRIPTION
**Pull request by hrsh7th**
I implemented drag drop feature.

But currently implementation is too buggy.

I want to help this problem...

### CODE REVIEW

File: debug/index.html [hunk 1]
```diff
@@ -39,6 +39,29 @@
         task.to = new Date(task.to.getTime() + 1000 * 60 * 60 * 24);
         roadmappy.render();
       });
+      roadmappy.on('drag:start:task:from', drag.bind(null, 'from'));
```
Comment (hrsh7th) [inline debug/index.html:42]: Handle events for debug.

File: src/Roadmappy.js [hunk 2]
```diff
@@ -17,15 +17,24 @@ export class Roadmappy extends EventEmitter {
     super();
 
     this.roadmap = new RoadmapParser().parse(dataSet);
-    this.canvas = new RoadmapCanvas(new RoadmapOption(options));
+    this.canvas = new RoadmapCanvas(this.roadmap, new RoadmapOption(options));
     this.canvas.on('click:task', this.emit.bind(this, 'click:task'));
+    this.canvas.on('drag:start:task', this.emit.bind(this, 'drag:start:task'));
```
Comment (hrsh7th) [inline src/Roadmappy.js:22]: Proxy canvas events.
It seems like reference.

File: src/Roadmappy.js [hunk 3]
```diff
@@ -17,15 +17,24 @@ export class Roadmappy extends EventEmitter {
...
+    this.canvas.on('drag:drag:task:to', this.emit.bind(this, 'drag:drag:task:to'));
+    this.canvas.on('drag:end:task:to', this.emit.bind(this, 'drag:end:task:to'));
   }
 
   /**
    * @returns {void}
    */
   render() {
-    this.canvas.render(this.roadmap);
+    this.canvas.render();
```
Comment (hrsh7th) [inline src/Roadmappy.js:37]: Cosmetic changes.
But i no satisfied with this code.
I will refactor this.

File: src/component/RoadmapCanvas.js [hunk 4]
```diff
@@ -5,117 +5,107 @@ import {AbstractRoadmapGroup} from './group/AbstractRoadmapGroup';
...
-    this._appendTaskLines(tasks, svg, xScale, yScale);
+    this._updateXScale(tasks, w);
+    this._updateYScale(tasks, h);
+    this._updateXAxis(w, h);
+    this._updateYAxis();
+    this._updateTaskBars(tasks);
 
     // utility.
-    this._addMouseHelper(roadmap, svg, xScale, yScale, marginLeft);
+    // this._addMouseHelper(marginLeft);
```
Comment (hrsh7th) [inline src/component/RoadmapCanvas.js:65]: Sorry...I was disable this feature.
When drag&drop, mouse helper dupulicate draw lines...

**Comment by hrsh7th** [comment]
I used this feature.
But it's very buggy.

Because, tasks are sorted from property, so task was moved at dragging. It's very no usual.

**Comment by atago0129** [comment]
It can be resolved to cache `sorted***`.
For example,
```
  _groupSortedTasks = {};

  /**
   * @param {string} type
   * @return {RoadmapTask[]}
   */
  getGroupSortedTasks(type) {
    if (!this._groupSortedTasks.hasOwnProperty(type)) {
      this._groupSortedTasks[type] = this.getSortedGroups(type).reduce((tasks, group) => {
        return tasks.concat(this.getSortedTasksByGroup(group));
      }, []);
    }
    return this._groupSortedTasks[type];
  }
```

But, it may not be good solution...

File: debug/index.html [hunk 5]
```diff
@@ -39,6 +39,29 @@
...
       });
+      roadmappy.on('drag:start:task:from', drag.bind(null, 'from'));
+      roadmappy.on('drag:drag:task:from', drag.bind(null, 'from'));
+      roadmappy.on('drag:end:task:from', drag.bind(null, 'from'));
+      roadmappy.on('drag:start:task:to', drag.bind(null, 'to'));
+      roadmappy.on('drag:drag:task:to', drag.bind(null, 'to'));
+      roadmappy.on('drag:end:task:to', drag.bind(null, 'to'));
+      function drag(key, task, pos) {
+        const xScale = roadmappy.canvas.xScale;
+        task[key] = xScale.invert(pos.x);
```
Comment (atago0129) [inline debug/index.html:50]: I would not like to support `time` format, and should validate `from` & `to`.
So I tried to fix to
```
        const targetDate = new Date(target.getFullYear() + '-' + (target.getMonth() + 1) + '-' + target.getDate());
        switch (key) {
          case 'to':
            if (task['from'] >= targetDate.setHours((new Date(targetDate).getHours() - 24))) return;
            break;
          case 'from':
            if (targetDate.setHours((new Date(targetDate).getHours() + 24)) >= task['to']) return;
            break;
        }
        task[key] = new Date(target.getFullYear() + '-' + (target.getMonth() + 1) + '-' + target.getDate());

```
but it sometimes has a bug that will be added 9 hour.

File: debug/index.html [hunk 6]
```diff
@@ -39,6 +39,29 @@
...
+        roadmappy.render();
+      }
+
+      roadmappy.on('drag:start:task', move);
+      roadmappy.on('drag:drag:task', move);
+      roadmappy.on('drag:end:task', move);
+      function move(task, pos) {
```
Comment (atago0129) [inline debug/index.html:57]: not work?

Comment (atago0129) [reply debug/index.html:57]: Sorry, it works.
I did not pinched a `task text`

**Comment by atago0129** [comment]
merge at once

**Pull request merged**